core/templates/services/math-interactions.service.ts

```diff
@@ -183,6 +183,102 @@ export class MathInteractionsService {
+    let shouldModifyNextTerm: boolean = false;
+    let modifyTerm = function(termString: string): string {
+      // If the shouldModifyNextTerm flag is set to true, we add the '-' sign,
+      // raise the term to a power of -1. This ensures that when the final
+      // list is joined by the '+'/'*' sign, it matches with the original
+      // expression. For eg.
+      // '3/10' would be split as [3, 10^(-1)] and
+      // '3-10' would be split as [3, -(10)].
+      if (splitByAddition) {
+        return '-(' + termString + ')';
+      } else {
+        return '(' + termString + ')^(-1)';
+      }
+    };
+    const primaryDelimiter = splitByAddition ? '+' : '*';
+    const secondaryDelimiter = splitByAddition ? '-' : '/';
+
+    expressionString = expressionString.replace(/\s/g, '');
+
+    // Temporarily replacing all unary negation signs with '~' so as to avoid
+    // splitting terms by them. We only need to split terms by binary
+    // subtraction signs and not unary negation signs. A '-' sign is considered
+    // to be binary subtraction iff it is preceded by an alphanumeric or a
+    // closing bracket, otherwise it is considered as unary negation operation.
+    // NOTE: The replace function is called twice to deal with cases where there
+    // might be overlapping matches.
+    // For eg. 4----5 would be converted to 4-~-~5 after the first call. So we
+    // need a second call to convert it to the desired result, which is 4-~~~5.
+    expressionString = expressionString.replace(/([^a-zA-Z0-9\)])-/g, '$1~');
+    expressionString = expressionString.replace(/([^a-zA-Z0-9\)])-/g, '$1~');
+
+    for (let i = 0; i < expressionString.length; i++) {
+      let currentVal = expressionString[i];
+      if (currentVal === '(' || currentVal === ')') {
+        bracketBalance += (currentVal === '(') ? 1 : -1;
+      }
+
+      // Split term only if we are not inside a set of parens and the current
+      // value is a delimiter.
+      if (bracketBalance === 0 && (
+        currentVal === primaryDelimiter || currentVal === secondaryDelimiter)) {
+        if (currentTerm.length !== 0) {
+          if (shouldModifyNextTerm) {
+            currentTerm = modifyTerm(currentTerm);
+            shouldModifyNextTerm = false;
+          }
+          listOfTerms.push(currentTerm);
+          currentTerm = '';
+        }
+        if (currentVal === secondaryDelimiter) {
+          shouldModifyNextTerm = true;
+        }
+      } else {
+        currentTerm += currentVal;
+      }
+    }
+    if (shouldModifyNextTerm) {
+      currentTerm = modifyTerm(currentTerm);
+      shouldModifyNextTerm = false;
+    }
+    listOfTerms.push(currentTerm);
+
+    // Reverting the temporary '~' replace in the final list of terms.
+    for (let i = 0; i < listOfTerms.length; i++) {
+      listOfTerms[i] = listOfTerms[i].replace(/~/g, '-');
+    }
+    return listOfTerms;
+  }
+
+  termsMatch(term1: string, term2: string): boolean {
```

Comment:
The  input to this should be terms  split by addition/subtraction, right? Add that as a comment.

Reply:
Done.